collectd/src/main/java/com/microsoft/applicationinsights/collectd/ApplicationInsightsWriter.java

```diff
@@ -0,0 +1,283 @@
+     */
+    protected ApplicationInsightsWriter(TelemetryClient telemetryClient, ApplicationInsightsWriterLogger logger) {
+        this.telemetryClient = telemetryClient;
+        this.telemetryConfiguration = TelemetryConfiguration.getActive();
+        this.logger = logger;
+    }
+
+    // endregion Ctors
+
+    // region Public
+
+    /**
+     * Reads the configuration provided for the writer plugin.
+     * @param configuration The configuration.
+     * @return Zero on success, non-zero otherwise. Non-zero values will disable the plugin entirely.
+     */
+    public int config(OConfigItem configuration) {
+        logger.logInfo("Loading configuration.");
+
+        WriterConfiguration writerConfiguration = null;
+        try {
+            writerConfiguration = WriterConfiguration.buildConfiguration(configuration);
+        } catch (ConfigurationException e) {
+            logger.logError("Configuration failed, plugin will be disabled:\n" + e.toString());
+
+            return CONFIGURATION_PHASE_ERROR_CODE;
+        }
+
+        this.telemetryConfiguration.setInstrumentationKey(writerConfiguration.getInstrumentationKey());
+        if (writerConfiguration.getIsLoggerEnabled()) {
+            InternalLogger.INSTANCE.initialize(DEFAULT_AI_LOGGER_OUTPUT, new HashMap<String, String>());
+        }
+
+        this.excludedPluginsDictionary = writerConfiguration.getPluginExclusions();
+        logger.logInfo("Configuration loaded successfully.");
+
+        return SUCCESS_CODE;
+    }
+
+    /**
+     * Initializes the writer plugin.
+     * @return Zero on success, non-zero otherwise. Non-zero values will disable the plugin entirely.
+     */
+    public int init() {
+        logger.logInfo("Initializing...");
+
+        try {
+            this.telemetryClient = new TelemetryClient(this.telemetryConfiguration);
+
+            logger.logInfo("Initialization completed.");
+        } catch (Throwable e ) {
+            logger.logError("Initialization failed, plugin will be disabled:\n" + e.toString());
+
+            return INITIALIZATION_PHASE_ERROR_CODE;
+        }
+
+        return SUCCESS_CODE;
+    }
+
+    /**
+     * Shutting down the writer plugin and flushes the telemetries from memory.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int shutdown() {
+        logger.logInfo("Flushing telemetries and shutting down...");
+        this.telemetryClient.flush();
+
+        return SUCCESS_CODE;
+    }
+
+    /**
+     * Writes the given values to Application Insights.
+     * @param valueList The values to be sent.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int write(ValueList valueList) {
+        // We're avoiding writing traces in this method to not spam the logs, as this method is being called very frequently.
+        logger.logDebug("Writing values (" + valueList.getValues().size() + "):\n" + valueList.toString());
+
+        String pluginName = valueList.getPlugin();
```

Comment:
If this method is being called very frequently won't it be best to store the data source to be included in a cache instead of iterating over all plugins and calling `isDataSourceExcluded` on them?